### ISSUE TITLE
[scaffolding-chef-infra] Primary policyfile lockfile being deleted when `include_policy`

### ISSUE THREAD

**Issue by echohack** [chef/effortless#41]
Originally from `habitat-sh/core-plans` by @qubitrenegade 

Hello,

I am having an issue with the `scaffolding-chef` scaffolding, where it appears to be deleting the `base.lockfile.json` when including the `base` policy from another policyfile.

I've created a [demo repo](https://github.com/qubitrenegade/habichef_demo).  The branch `2019-04-28-pt1-multiple-policy-files` has been created with these changes.

I have a fairly standard [plan.sh](https://github.com/qubitrenegade/habichef_demo/blob/2019-04-28-pt1-multiple-policy-files/habitat/plan.sh) file.

And my policy file is:

```
# habichef_demo/policyfiles/us-east-1.rb
name 'us-east-1'
include_policy 'base', path: '.'
default['resolver']['domain'] = 'us-east-1.qubitrenegade.com'
```

When running a build, this fails with:

```
[2][default:/src:0]# CHEF_POLICYFILE=us-east-1 build
... snip ...
Building policy us-east-1
Error: Failed to generate Policyfile.lock
Reason: (ChefDK::LocalPolicyfileLockNotFound) The provided path ./base.lock.json does not exist.
```

If I build it first with:

```
[1][default:/src:0]# build
... snip ...

  habichef_demo-base: Installed Path: /hab/pkgs/qbrd/habichef_demo-base/0.1.0/20190428185221
   habichef_demo-base: Artifact: /src/results/qbrd-habichef_demo-base-0.1.0-20190428185221-x86_64-linux.hart
   habichef_demo-base: Build Report: /src/results/last_build.env
   habichef_demo-base: SHA256 Checksum: 4a126ec5c47dba1643ef767c7ccc02acd36ea780d2f88110db9c964c4598c2b9
   habichef_demo-base: Blake2b Checksum: 7fc45fcf07bebe57f90301c0c83df6700f772fc8925ec0162b1efbfa85b89b6a
   habichef_demo-base:
   habichef_demo-base: I love it when a plan.sh comes together.
   habichef_demo-base:
   habichef_demo-base: Build time: 0m5s
[2][default:/src:0]#
```

It creates a `base.lockfile.json`, but running `CHEF_POLICYFILE=us-east-1 build` again fails with the same error.

This appears to be the result of PR #2515, [L99](https://github.com/habitat-sh/core-plans/blame/932bfaae647d3bd9dde892158b7077b1d19888fc/scaffolding-chef/lib/scaffolding.sh#L99)

```
 rm -f "${scaffold_policyfiles_path}"/*.lock.json
```

Thanks
- QBRD

**Comment by echohack** [comment]
Hi @qubitrenegade Thank you for opening this issue.

I don't think the ability to inherit from multiple policyfiles has been supported by the scaffolding so far. I can see us adding a feature to change the current single policyfile approach to an array of policyfiles that get compiled... although this would probably have to be ordered correctly.

`rm -f "${scaffold_policyfiles_path}"/*.lock.json` has been present since the start of the scaffolding. It's possible we could change this to only delete the `${scaffold_policy_name}.lock.json` as a short term fix.

I have a few questions:


1. Are you checking in your `base.lock.json` policyfile into source control?
2. **When** are you expecting `base.lock.json` to be built? In the build for `us-east-1.rb` policyfile build?

**Comment by echohack** [comment]
> I don't think the ability to inherit from multiple policyfiles has been supported by the scaffolding so far.

Oh!  Well I blame @smford22 then!  :grinning: 

> I can see us adding a feature to change the current single policyfile approach to an array of policyfiles that get compiled...

Ok, well then I guess this is a feature request! :wink: 

> `rm -f "${scaffold_policyfiles_path}"/*.lock.json` has been present since the start of the scaffolding

Admittedly I only looked at the `git blame` and didn't dig too far back.  I could have sworn I had built multiple policy files in the past, but I fail to find my example code... so I _could_ be imagining things.

> 1. Are you checking in your base.lock.json policyfile into source control?

Yes...?  But only because there isn't a generator for the `chef-scaffolding` and therefore I didn't think to add it to the `.gitignore`.

I'm still kinda new to working with Policyfiles.  Is that common?

> 2. When are you expecting base.lock.json to be built? In the build for us-east-1.rb policyfile build?

Well...  that's a good question.  [The way I was imagining](https://github.com/qubitrenegade/qubitrenegade.github.io/blob/master/_posts/2019-04-28-Habichef-Pt2-Applying-Change.md#multiple-policyfiles) (and perhaps assuming) a multiple policy-file workflow to work, would be `build base, build policy that includes base, build policy that includes policy that includes base, etc.`.

E.g.:

1. `CHEF_POLICYFILE=base hab pkg build .`
2. `CHEF_POLICYFILE=us-east-1 hab pkg build .`

But actually, it might be nice to go "just build all my policyfiles" with like a `CHEF_POLICYFILE=all hab pkg build .` and have Habitat spit out 10 packages (if I had 10 policy files).

I guess you raise a good point, does Chef have a way for resolving policy file dependencies?  (I don't think it does, hence the error "no base.lock.json" instead of compiling it...)  If not, does that logic really belong in Habitat or would it maybe be more appropriate to open a Chef Feature Request for "Policyfile dependency resolution" or so.  (i.e.: "when compiling `us-east-1`, `chef` knows that it needs to compile `base` first if no `base.lockfile.json` exits")

Thanks for the quick reply!

**Comment by TheLunaticScripter** [comment]
Just to give a quick update here for this issue. I went back in my time machine and in the past we were able to do this. The way it worked is we would build the included policy first then build the new policyfile. The issue is in this [for](https://github.com/chef/effortless/blob/fcb877b39262b450ee6c00a28c8974239cb32400/scaffolding-chef-infra/lib/scaffolding.sh#L120) statement. Looking at the old code I think this needs to be something like this.
```bash
for p in `cat $policyfile | grep include_policy | awk -F "," '{print $1}' | awk -F '"' '{print $2}' | tr -d " "`; do
  chef install $scaffold_policyfile_path/$p.rb
done
```

We are currently in the middle of a file re-write but as soon as that gets in I'll take a look at this.

**Comment by TheLunaticScripter** [comment]
I started working on this today. Turns out that the above suggestion has nothing to do with the problem. The issue is that the pipe to `awk -f '"' '{print $2}'` is looking for what is between a double quote. In the demo repo line 6 is `include_policy 'base', path './base.lock.json'`. This is a single quote and so the `for p` statement return nothing. So the fix to this is to change line 6 to `include_policy "base", path './base.lock.json'`. I'm going to add a case statement to this to also look for a single quote which should fix this.
